File: google-adk/inis_google_adk/tools.py
"""Explicit, non-experimental fallback: plain `google.adk.tools.function_tool.
FunctionTool` wrappers around an inis.run session, for callers who don't want
to depend on ADK's `@experimental` `BaseEnvironment`/`EnvironmentToolset`
surface at all (see `environment.py`'s module docstring for exactly what's
experimental and why). `google.adk.tools.function_tool.FunctionTool` itself
carries no `@experimental` decorator in installed source -- verified, not
assumed -- so this path is the stable one per the issue's "provide an
explicit toolset fallback while the environment API remains experimental"
requirement.

Mirrors `inis_pydantic_ai.toolset.inis_toolset()`'s shape: the smallest
execution/file surface a model needs, nothing that creates, destroys,
pauses, or otherwise administers sessions. Attach-only -- pass an existing
`inis.AsyncSession`'s ID; this module never creates or destroys one.
"""

from __future__ import annotations

import secrets

from google.adk.tools.function_tool import FunctionTool

from inis import AsyncSession, InisError


def inis_adk_tools(session: AsyncSession) -> list[FunctionTool]:
    """Return plain ADK `FunctionTool`s (`execute`, `read_file`, `write_file`,
    `list_files`) bound to an already-created/attached inis.run session.

    Unlike `InisEnvironment`/`EnvironmentToolset`, these are not experimental
    upstream and do not require ADK's environment surface at all -- pass the
    result straight to `Agent(tools=[...])`. This adapter never creates or
    destroys `session`; the host owns that, one session per agent thread,
    same as everywhere else in this package.
    """
    if not session.session_id:
        raise ValueError("inis_adk_tools() requires an already-created/attached session")

    async def execute(command: str) -> str:
        """Run a shell command in the inis.run session and return its output.

        Args:
          command: The shell command to run. Chain dependent commands with &&.
        """
        result = await session.exec(command)
        output = result.stdout
        if result.stderr:
            output = f"{output}\n[stderr]\n{result.stderr}" if output else result.stderr
        if result.timed_out:
            output = f"{output}\n[inis: command timed out]" if output else "[inis: command timed out]"
        if result.exit_code != 0:
            output = f"{output}\n[inis: exited {result.exit_code}]"
        if result.truncated:
            # ExecResult.truncated is real (sdk/python/inis/client.py) but
            # this function returns a single string with no separate field
            # to carry it -- append a visible marker instead of silently
            # dropping the fact that what's above is a partial view, the
            # same choice InisEnvironment.execute() makes for the identical
            # gap (see environment.py). Nonce is fresh per call: a guest
            # with full control of its own output could otherwise read this
            # marker's exact wording out of this open-source file and print
            # a fake one.
            nonce = secrets.token_hex(8)
            marker = f"[inis: output truncated by the sandbox's own capture limit (nonce={nonce})]"
            output = f"{output}\n{marker}" if output else marker
        return output

    async def read_file(path: str) -> str:
        """Read a file from the inis.run session's filesystem.

        Args:
          path: Absolute or /workspace-relative path to the file.
        """
        try:
            return await session.read_file(path)
        except InisError as e:
            return f"[inis: read failed: {e}]"

    async def write_file(path: str, content: str) -> str:
        """Write a file to the inis.run session's filesystem.

        Args:
          path: Absolute or /workspace-relative path to the file.
          content: The full text content to write.
        """
        await session.write_file(path, content)
        return f"wrote {len(content)} bytes to {path}"

    async def list_files(path: str = "/workspace") -> str:
        """List files in a directory in the inis.run session (defaults to /workspace).

        Args:
          path: Absolute directory path to list inside the session.
        """
        entries = await session.list_files(path)
        return "\n".join(entries)

    return [
        FunctionTool(execute),
        FunctionTool(read_file),
        FunctionTool(write_file),
        FunctionTool(list_files),
    ]

File: ai-sdk/src/tools.ts
import { tool } from "ai";
import { Client, InisError, Session, type ConnectionCreate } from "@inis-run/sdk";
import { z } from "zod";

export interface InisClientOptions {
  /** API key; defaults to $INIS_API_KEY */
  apiKey?: string;
  /** API base URL; defaults to $INIS_BASE_URL or https://api.inis.run */
  baseUrl?: string;
  timeoutMs?: number;
}

export interface InisToolsOptions extends InisClientOptions {
  /**
   * ID of a session your app already created or attached to. Required:
   * inisTools() only ever attaches, never creates or discovers a session on
   * its own: one agent conversation maps to one session, with no global
   * mutable session state. Never destroyed by this call — see
   * `createOwnedInisTools()` if you want the tools package to own creating
   * (and destroying) the session for you.
   */
  sessionId: string;
}

type InisToolSet = ReturnType<typeof buildExecutionTools>;

/** The bounded execution/file surface shared by inisTools() and
 * createOwnedInisTools() — built directly against an already-open Session,
 * so callers that already have one (the owned path) don't re-attach. */
function buildExecutionTools(session: Session) {
  return {
    execute_python: tool({
      description: "Run Python code in the sandbox and return stdout.",
      inputSchema: z.object({
        code: z.string().describe("Python source code to execute"),
      }),
      execute: async ({ code }) => {
        const result = await session.exec(["python3", "-c", code]);
        if (result.exitCode !== 0) {
          throw new InisError(
            `python exited ${result.exitCode}: ${result.stderr || result.stdout}`,
          );
        }
        if (result.stderr) {
          return `${result.stdout}\n[stderr]\n${result.stderr}`.trim();
        }
        return result.stdout;
      },
    }),

    execute_shell: tool({
      description: "Run a shell command in the sandbox and return stdout.",
      inputSchema: z.object({
        command: z.string().describe("Shell command to run"),
      }),
      execute: async ({ command }) => {
        const result = await session.exec(command);
        if (result.exitCode !== 0) {
          throw new InisError(
            `command failed (exit ${result.exitCode}): ${result.stderr || result.stdout}`,
          );
        }
        return result.stdout;
      },
    }),

    read_file: tool({
      description: "Read a file from /workspace in the sandbox.",
      inputSchema: z.object({
        path: z.string().describe("Absolute or workspace-relative file path"),
      }),
      execute: async ({ path }) => session.readFile(path),
    }),

    write_file: tool({
      description: "Write a file under /workspace in the sandbox.",
      inputSchema: z.object({
        path: z.string().describe("Absolute or workspace-relative file path"),
        content: z.string().describe("File contents to write"),
      }),
      execute: async ({ path, content }) => {
        await session.writeFile(path, content);
        return `wrote ${content.length} bytes to ${path}`;
      },
    }),

    list_files: tool({
      description:
        "List the files and directories inside a path in the sandbox. " +
        "Defaults to /workspace. Returns a newline-separated list of entry names.",
      inputSchema: z.object({
        path: z
          .string()
          .default("/workspace")
          .describe("Absolute directory path to list inside the sandbox (defaults to /workspace)"),
      }),
      execute: async ({ path }) => {
        const entries = await session.listFiles(path);
        return entries.join("\n");
      },
    }),
  };
}

function makeClient(opts: InisClientOptions): Client {
  return new Client({ token: opts.apiKey, baseUrl: opts.baseUrl, timeoutMs: opts.timeoutMs });
}

/**
 * The smallest execution/file surface a model needs: execute_python,
 * execute_shell, read_file, write_file, list_files. Attach-only — pass the
 * ID of a session your app already owns or is attached to. Nothing here
 * creates, destroys, pauses, resumes, or exposes a session — the
 * model-facing default exposes only the bounded execution/file surface it
 * needs, while credentials and lifecycle administration remain
 * host-controlled or explicitly opt-in. For session/sandbox administration
 * or one-shot sandboxes, see `inisLifecycleTools()` / `inisEphemeralTools()` in
 * `@inis-run/ai-sdk/lifecycle` — deliberate opt-ins, not exported here.
 */
export function inisTools(opts: InisToolsOptions): InisToolSet {
  if (!opts.sessionId) {
    throw new InisError(
      "inisTools() requires sessionId: attach to a session your app already " +
        "created or is attached to. Use createOwnedInisTools() if you want " +
        "this call to create (and later destroy) the session for you.",
    );
  }
  const client = makeClient(opts);
  const session = Session.attach(client, opts.sessionId);
  return buildExecutionTools(session);
}

export interface OwnedInisTools {
  /** The same bounded execution/file tools inisTools() returns. */
  tools: InisToolSet;
  /** ID of the session this call created, for reconnecting on a later request. */
  sessionId: string;
  /** Destroys the session this call created. Call exactly once, when the
   * conversation is over — success or failure. */
  cleanup: () => Promise<void>;
}

/**
 * Deliberate owned-session factory: creates a NEW session, returns its
 * tools alongside the session ID and a cleanup handle that destroys it.
 * Unlike inisTools(), this call does create (and, via `cleanup`, destroy) a
 * session — that's the entire point, so it's a distinct, explicitly-named
 * function rather than an implicit "sessionId omitted -> auto-create" path
 * (the previous shape here claimed a null sessionRef auto-created when the
 * code actually threw).
 *
 * One call here = one owned session for one conversation. Do not share the
 * returned `tools`/`sessionId` across concurrent conversations — construct
 * a new one per conversation. There is no global mutable session state.
 */
export async function createOwnedInisTools(
  opts: InisClientOptions & {
    egress?: "allow" | "deny";
    /**
     * Inline Connections to create with the session: session-scoped
     * credential leases bound to specific external API origins. See
     * `ConnectionCreate` in `@inis-run/sdk` for the full shape.
     */
    connections?: ConnectionCreate[];
  } = {},
): Promise<OwnedInisTools> {
  const client = makeClient(opts);
  const createOpts: { egress?: { mode: "allow" | "deny" }; connections?: ConnectionCreate[] } = {};
  if (opts.egress) createOpts.egress = { mode: opts.egress };
  if (opts.connections) createOpts.connections = opts.connections;
  const session = await client.sessions.create(
    Object.keys(createOpts).length > 0 ? createOpts : undefined,
  );
  return {
    tools: buildExecutionTools(session),
    sessionId: session.sessionId,
    cleanup: () => session.destroy(),
  };
}

export { buildExecutionTools as _buildExecutionTools, makeClient as _makeClient };
